### ISSUE TITLE
Split multi-valued cells should respect facets

### ISSUE THREAD

**Issue by benmiller314** [OpenRefine/OpenRefine#5978]
In my dataset, I have several records that include appropriately multi-valued cells, with values all delimited by " | ". For just one of those records, though, I have a reason to split it out into two single-valued records, each with their own row. I've faceted so that this row is the only one that matches, but when I do "split multi-valued cells," OpenRefine seems to be splitting values elsewhere, too.

Am I misunderstanding the scope of this operation? I had thought all operations only acted on rows matching the current filters and facets.

### To Reproduce
1. Import a dataset with multi-valued cells:

<pre>record_id,facet_on_me,multi_value,other_values
1,red,Monday | Tuesday,fire
2,blue,Wednesday,water
3,yellow,Thursday | Friday,electric
4,brown,Saturday,earth</pre>

2. Facet on the second column and select "red"; only one record (and one row) will match.

3. Split the third column on " | ". The record will now have two rows, as expected.

4. Clear the facet. Record 3 now also has two rows, despite not having matched the facet selected during the operation.

### Current Results

All multivalues in the column are split, including in records that hadn't been visible.

<img width="494" alt="Screen Shot 2023-07-17 at 12 43 06 PM" src="https://github.com/OpenRefine/OpenRefine/assets/3182049/05d30d6e-b677-4367-a859-c3564aab8f4f">

### Expected Behavior

I'd expected that I was only editing the visible records; that would have led to something like this:

<img width="477" alt="Screen Shot 2023-07-17 at 12 44 30 PM" src="https://github.com/OpenRefine/OpenRefine/assets/3182049/c673cbe5-e854-4e81-810f-e74900d9a906">

### Screenshots

![where do the other four rows come from](https://github.com/OpenRefine/OpenRefine/assets/3182049/c45eb024-c835-4e6a-b571-fc549ddef549)

### Versions
 - Operating System:  MacOS Monterey 12.6.7
 - Browser Version:  Chrome 114.0.5735.198 (Official Build) (arm64)
 - JRE or JDK Version:  Java(TM) SE Runtime Environment (build 18.0.1+10-24)
 - OpenRefine:  Version 3.7.3 [22b54c4]

### Discussion
I have also [posted this question to the community forum](https://forum.openrefine.org/t/split-multi-valued-cells-seems-to-affect-non-matching-rows/870); if that's the more appropriate place to discuss, I'll continue there instead, but as of the present moment there were no replies on that thread.

If I've misunderstood the scope and Split Multi-Valued Cells is working as intended, may I request adding a note to that effect in the [documentation for this operation](https://openrefine.org/docs/manual/cellediting#split-multi-valued-cells)? I've searched and couldn't find any indication that it would always affect the entire column, regardless of facets, and I believe that's different from how other operations work.

**Comment by dauinh** [comment]
hi @tfmorris @wetneb I'm a GSoC applicant and I want to understand OpenRefine better. Can I try with this issue?

**Comment by wetneb** [comment]
@dauinh sure! What would be your approach to it? Do you want to change the behavior of the operation so that it respects facets, or rather document the existing behavior?

**Comment by dauinh** [comment]
@wetneb Can I try with changing the behavior? 

But for me to understand this issue better, could you explain more about a couple of things mentioned before? Please direct me to any documentation that I can refer to!

- I don't quite understand the concept of 'facet', what does it mean exactly? How do you respect facets? 

- What does the facetted engine do?

> sometimes putting the split value into an existing row and other times adding a new row to put it in, it requires access to all the rows in the project, which is not possible using the row visitor pattern which is used by the facetted engine.

Thank you!

**Comment by wetneb** [comment]
Yes, I think changing the operations' behaviour would be sensible.

If you are not familiar with facets yet, that's a sign that it could be worth following [our guide to first pull requests](https://openrefine.org/docs/technical-reference/code-contributions), which recommends you to get familiar with the tool as a user first. It points you to a few tutorials that could be good starting points. Otherwise, you could also find others by googling something like "openrefine facet tutorial" and pick one of the many results.

**Comment by tfmorris** [comment]
I've reverted the title and labels given the change in focus. Documentation is now in #6413

**Comment by dauinh** [comment]
After going through tutorial, documentation and codebase, I have understood the issue a bit more. But I might need some guidance as I'm very new to the codebase. My proposed solution is to save resulting rows from facet operation `getFilteredRows()` to a new class variable in `Engine.java`. Then for all other operations, use this variable. If it's empty, then use all filtered rows.

Please kindly suggest your opinions. Thank you!

**Comment by wetneb** [comment]
I don't think you should need to make any changes to the `Engine` class. You should be able to process the filtered rows directly in the Operation class, where you would generate the appropriate change data. Why do you feel the need to store things in the `Engine` class itself?

**Comment by tfmorris** [comment]
When I adjusted the labels, I also removed the "Good First Issue" label. @dauinh if you were being guided by that label, you may want to reconsider. From an implementation point of view, this is probably underspecified for a new contributor.

Because the new implementation will have different behavior, a different JSON serialization (which affects the REST API), and a different operation history serialization, I would like to suggest that we introduce it as a completely new operation and retain the old operation for compatibility.

Also, since I suspect Split & Join to behave similarly, I would suggest that new versions of both operations be created at the same time. If they are new operations, with new names, as I suggest, front end modifications will be needed to invoke the new operations (and send the necessary facets configuration). Not hugely complicated, but one more thing to address.

**Comment by tfmorris** [comment]
One other wrinkle is that the current behavior will reuse empty cells in the current record to split into in lieu of adding a new row. We need to decide if the new operation should continue that behavior or not and also what to do if that row is filtered out by facets. Also, with the Visitor pattern that's used we don't actually have visibility into whether or not a future row will be filtered, so it probably would be necessary to cache the newly split values and merge them into future rows when they are visited (or create new rows for them later).

I haven't seen anything that's impossible to overcome, but I can definitely see why David implemented it the way he did.

**Comment by dauinh** [comment]
Thank you both for your feedback. I must admit that I'm a bit nervous tackling this issue, but I want to try if I can handle the challenge. I'm new to open sources and I really appreciate any opportunity to learn and contribute. Could I follow this issue and help out in any way? Or is it a better idea to move on to other issues?

**Comment by wetneb** [comment]
I would recommend you tackle another issue.
